### PR TITLE
[dv/csr_utils] update unmapped_addr calculation

### PR DESCRIPTION
**Pull request by udinator**
Addresses #6594 and #6596.

This PR is split into 2 commits.

The first commit overhauls how we calculate unmapped address ranges for a given reg block.
We now compute a list of mapped/unmapped address ranges and store them in the `dv_base_reg_block`
during the build phase.
Along with this, we move the `get_csr_addrs` and `get_mem_addr_ranges` functions into the `dv_base_reg_block`,
as this seems a more sensible place to put these methods.

The second commit updates the scoreboards for all IPs and the scoreboard template to use the
updated locations of `csr_addrs` and `mem_ranges` lists.

Signed-off-by: Udi Jonnalagadda <udij@google.com>

### CODE REVIEW

File: hw/dv/sv/csr_utils/csr_utils_pkg.sv [hunk 1]
```diff
@@ -82,6 +82,106 @@ package csr_utils_pkg;
...
+    uvm_reg_addr_t full_addr_range[int];
+
+    uvm_reg_addr_t csr_addrs[$];
+
+    // A queue of all of the currently mapped address ranges contained within the RAL model.
+    // Memory ranges are directly copied over to this list, but CSR addresses are converted
+    // into address ranges, as each CSR technically covers a range of
+    // `[csr_address : csr_address + csr.get_n_bytes() - 1]`.
+    addr_range_t   addr_ranges_to_remove[$];
+
```
Comment (matutem) [inline hw/dv/sv/csr_utils/csr_utils_pkg.sv:107]: Extra line

File: hw/dv/sv/dv_lib/dv_base_env_cfg.sv [hunk 2]
```diff
@@ -41,6 +41,12 @@ class dv_base_env_cfg #(type RAL_T = dv_base_reg_block) extends uvm_object;
   bit [bus_params_pkg::BUS_AW-1:0]  csr_addrs[string][$];
   addr_range_t                      mem_ranges[string][$];
 
+  // Associative array of bits, each one corresponding to an instantiated RAL model and indicating
+  // whether the RAL model has unmapped addresses.
+  bit ral_has_unmapped_addrs[string];
```
Comment (sriyerg) [inline hw/dv/sv/dv_lib/dv_base_env_cfg.sv:46]: Not really sure if this is needed. `unmapped_addr_ranges["foo"].size()` already provides this information.

File: hw/dv/sv/csr_utils/csr_utils_pkg.sv [hunk 3]
```diff
@@ -82,6 +82,106 @@ package csr_utils_pkg;
...
   endfunction
 
+  // For internal use only.
+  //
+  // This function scans all registers/memories contained within a dv_base_reg_block RAL,
+  // and produces a queue of the unmapped address ranges as `addr_range_t` structs.
+  //
+  // As this function can be very expensive, it is only called once when the RAL model has just been
+  // built.
+  function automatic void compute_unmapped_addr_ranges(dv_base_reg_block ral,
```
Comment (sriyerg) [inline hw/dv/sv/csr_utils/csr_utils_pkg.sv:92]: I thought we decided that this belongs to `dv_base_reg_block`?

Comment (udinator) [reply hw/dv/sv/csr_utils/csr_utils_pkg.sv:92]: I decided to change to `csr_utils_pkg` to be able to take advantage of `get_csr_addrs` and `get_mem_ranges` methods, rather than duplicate the method bodies if I implemented this inside `dv_base_reg_block`.

File: hw/dv/sv/dv_lib/dv_base_env_cfg.sv (hunk 2 above)
Comment (udinator) [reply hw/dv/sv/dv_lib/dv_base_env_cfg.sv:46]: That's true...looking at it now it's not really needed at all. Let me get rid of it, thanks!

File: hw/dv/sv/csr_utils/csr_utils_pkg.sv (hunk 1 above)
Comment (udinator) [reply hw/dv/sv/csr_utils/csr_utils_pkg.sv:107]: will remove, thanks!

File: hw/dv/sv/csr_utils/csr_utils_pkg.sv [hunk 4]
```diff
@@ -82,6 +82,106 @@ package csr_utils_pkg;
...
+
+    uvm_reg_addr_t full_addr_range[int];
+
+    uvm_reg_addr_t csr_addrs[$];
+
+    // A queue of all of the currently mapped address ranges contained within the RAL model.
+    // Memory ranges are directly copied over to this list, but CSR addresses are converted
+    // into address ranges, as each CSR technically covers a range of
+    // `[csr_address : csr_address + csr.get_n_bytes() - 1]`.
+    addr_range_t   addr_ranges_to_remove[$];
```
Comment (matutem) [inline hw/dv/sv/csr_utils/csr_utils_pkg.sv:106]: Using full_addr_range seems overkill. If you traverse the addr_ranges_to_remove, any pair of non-abutting ranges give rise to an unmapped range, with special cases for the start (base_addr) and last (base_addr + addr_mask).

File: hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh [hunk 5]
```diff
@@ -25,26 +25,21 @@
...
-  end
+  addr_range_t loc_unmapped_addr_ranges[$] = updated_unmapped_addr_ranges[ral_name];
 
   // randomize unmapped_addr first to improve perf
   repeat ($urandom_range(10, 100)) begin
     bit [BUS_AW-1:0] unmapped_addr;
-    addr_range_t loc_mem_ranges[$] = updated_mem_ranges[ral_name];
+
+    // Randomly pick which unmapped address range to target
+    int idx = $urandom_range(0, loc_unmapped_addr_ranges.size()-1);
```
Comment (matutem) [inline hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh:35]: What happens when there are no unmapped addresses? Seems like idx becomes $urandom_range(0, -1), and -1 is treated as unsigned, so it is any int, which most likely will result in an out of bounds address in line 40.

File: hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh [hunk 6]
```diff
@@ -197,11 +198,13 @@ virtual task run_tl_errors_vseq_sub(int num_times = 1, bit do_wait_clk = 0, stri
           fork
             begin
               randcase
-                1: tl_write_csr_word_unaligned_addr(ral_name);
```
Comment (matutem) [inline hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh:200]: Why is this removed?

File: hw/dv/sv/csr_utils/csr_utils_pkg.sv (hunk 4 above)
Comment (udinator) [reply hw/dv/sv/csr_utils/csr_utils_pkg.sv:106]: hmm, that is true. let me try rewriting this part

File: hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh [hunk 7]
```diff
@@ -197,11 +198,13 @@ virtual task run_tl_errors_vseq_sub(int num_times = 1, bit do_wait_clk = 0, stri
...
                 1: tl_write_less_than_csr_width(ral_name);
                 1: tl_protocol_err(p_sequencer.tl_sequencer_hs[ral_name]);
+
                 // only run when unmapped addr exists
-                has_unmapped_addr: tl_access_unmapped_addr(ral_name);
+                cfg.ral_has_unmapped_addrs[ral_name]: tl_access_unmapped_addr(ral_name);
+                cfg.ral_has_unmapped_addrs[ral_name]: tl_write_csr_word_unaligned_addr(ral_name);
```
Comment (matutem) [inline hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh:206]: Why generate unaligned addresses only if there are unmapped addresses?

File: hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh (hunk 5 above)
Comment (udinator) [reply hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh:35]: this actually will never be the case here - if you look at line 207 of this file, this task is only allowed to be invoked if unmapped addresses exist

File: hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh [hunk 8]
```diff
@@ -182,6 +175,14 @@ virtual task run_tl_errors_vseq_sub(int num_times = 1, bit do_wait_clk = 0, stri
     end
   end
 
+  if (cfg.ral_has_unmapped_addrs[ral_name]) begin
+    foreach (cfg.unmapped_addr_ranges[ral_name][i]) begin
+      updated_unmapped_addr_ranges[ral_name].push_back(addr_range_t'{
```
Comment (matutem) [inline hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh:180]: Seems like this is just the corresponding cfg.unmapped_addr_ranges & addr_mask. When you use it in line 39 you are masking it... it seems you could simply use unmapped_addr_ranges itself, i.e. this seems redundant.

File: hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh (hunk 5 above)
Comment (matutem) [reply hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh:35]: To be safe, could you just `return` if there are no unmapped ranges?

File: hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh (hunk 7 above)
Comment (udinator) [reply hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh]: hmm thinking about it more, perhaps what this needs to be instead is a check that the RAL model has CSR addresses (since this method constrains an address to not be within any memory range)

File: hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh (hunk 5 above)
Comment (udinator) [reply hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh:35]: done.

File: hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh (hunk 7 above)
Comment (udinator) [reply hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh]: done.

File: hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh (hunk 8 above)
Comment (udinator) [reply hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh:186]: good catch, thanks for pointing that out! will fix.

Comment (udinator) [reply hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh:186]: sorry - please disregard my previous comment. the `updated_unmapped_addr_ranges` is actually necessary, because we use these normalized addresses to create upper and lower bounds in the constraints of `tl_access_unmapped_addrs()`.

If we don't normalize these address ranges, then we have to do a last minute hack to either get the csr_base_addr again in the method or something else like that.

File: hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh [hunk 9]
```diff
@@ -164,12 +161,14 @@ virtual task run_tl_errors_vseq_sub(int num_times = 1, bit do_wait_clk = 0, stri
...
-  bit has_unmapped_addr;
+
+  bit has_csr_addrs = (cfg.csr_addrs[ral_name].size() > 0);
+
+  bit has_unmapped_addrs = (cfg.unmapped_addr_ranges[ral_name].size() > 0);
 
   // get_addr_mask returns address map size - 1 and get_max_offset return the offset of high byte
   // in address map. The difference btw them is unmapped address
   csr_addr_mask[ral_name] = cfg.ral_models[ral_name].get_addr_mask();
-  has_unmapped_addr = csr_addr_mask[ral_name] > cfg.ral_models[ral_name].get_max_offset();
```
Comment (weicaiyang) [inline hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh:172]: Thanks for finding and fixing these issues.
the fix for `has_csr_addrs` looks great

and I find a relative easy way to check if there is any unmapped address

```
uvm_reg regs[$];
uvm_mem mems[$];
get_memories(mems);
get_registers(regs);
// totally bytes of all mems
foreach (mems[i]) mem_bytes += mems.get_n_bytes * mems[i].get_n_bytes - 1;

csr_bytes = regs.size >> 2;

// if all the addresses are used, csr_bytes + mem_bytes - 1 should be equal to the mask
has_unmapped_addr = (csr_bytes + mem_bytes - 1) != get_addr_mask;

```

perhaps we can add this function in dv_base_reg_block

**Comment by tomroberts-lowrisc** [comment]
I've been looking at similar things while adding the rom_ctrl RAL model. Have you also seen issues with the `tl_protocol_err` sequence?
It looks like it drives completely unconstrained addresses which then mismatch with the rtl on `is_tl_access_mapped_addr`

**Comment by weicaiyang** [comment]
> I've been looking at similar things while adding the rom_ctrl RAL model. Have you also seen issues with the `tl_protocol_err` sequence?
> It looks like it drives completely unconstrained addresses which then mismatch with the rtl on `is_tl_access_mapped_addr`

@tomroberts-lowrisc can you elaborate the issue? 
does it happen at the test `tl_errors`? If so, that test will trigger all kinds of error cases, it never drives non-error cases (all transaction should have d_error = 1). If you see unmapped addresses are sent from that test, that could be expected.

**Comment by tomroberts-lowrisc** [comment]
Hi @weicaiyang (welcome back!) - yes I see cases where the address is unmapped, but `d_error == 1'b0` from that sequence. It's something to do with having two RAL models in the design as I didn't see it before.

I'll wait for the fixes from @udinator to go in, then I'll push my dual RAL changes as a draft PR to give an example

**Comment by weicaiyang** [comment]
> Hi @weicaiyang (welcome back!) - yes I see cases where the address is unmapped, but `d_error == 1'b0` from that sequence. It's something to do with having two RAL models in the design as I didn't see it before.
> 
> I'll wait for the fixes from @udinator to go in, then I'll push my dual RAL changes as a draft PR to give an example

I see. Thanks for the explanation.

**Comment by tomroberts-lowrisc** [comment]
Update - I tracked down my issue to traffic being driven on the wrong interface - fixed as part of #6830 
With that fix and the changes from this PR, everything looks good on the tl_error sequences for me

File: hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh (hunk 9 above)
Comment (udinator) [reply hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh:172]: thanks for the suggestion @weicaiyang !
I'm working on moving the `compute_unmapped_addr_ranges()` function into the `dv_base_reg_block`, so maybe we can keep the same calculation, but also move this bit into the base reg block?

Comment (weicaiyang) [reply hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh:172]: agree, sgtm. Thanks

File: hw/dv/sv/dv_base_reg/dv_base_reg_block.sv [hunk 10]
```diff
@@ -94,6 +103,95 @@ class dv_base_reg_block extends uvm_reg_block;
...
+
+      // end_addr+1 -> highest_addr
+      if (mapped_addr_ranges[$].end_addr < highest_addr) begin
+        range.start_addr = mapped_addr_ranges[$].end_addr + 1;
+        range.end_addr   = highest_addr;
+        unmapped_addr_ranges.push_back(range);
+      end
+    end
+    has_unmapped_addrs = (unmapped_addr_ranges.size() > 0);
+  endfunction
```
Comment (weicaiyang) [inline hw/dv/sv/dv_base_reg/dv_base_reg_block.sv:193]: it's a bit expense to maintain `mapped_addr_ranges` and `unmapped_addr_ranges`, but it will help the randomization and easy for debug if needed.
I'm thinking of a few way to make it less expense but with other costs
1. assume in each IP (each sub reg_block), CSR address is continuous. the range of csr_base_addr to csr_base_addr + num_of_csr * 4 -1. so each block will only have very a few items in mapped/unmapped_addr_ranges
2. For tl_error vseq, we only need `has_unmapped_addrs` and `mapped_addr_ranges`. we can maintain only 1 `mapped_addr_ranges`, find a simple way to calculate `has_unmapped_addrs`.
```
get_memories(mems); 
get_registers(regs); 
foreach (mems[i]) mem_bytes += mems.get_n_bytes * mems[i].get_n_bytes - 1; 
csr_bytes = regs.size >> 2; 
// if all the addresses are used, csr_bytes + mem_bytes - 1 should be equal to the mask 
has_unmapped_addr = (csr_bytes + mem_bytes - 1) < get_addr_mask;
```

I'm fine with current approach too, since the other two approaches aren't perfect. Maybe others can chime in. @sriyerg @matutem

File: hw/dv/sv/dv_base_reg/dv_base_reg_block.sv [hunk 11]
```diff
@@ -94,6 +103,95 @@ class dv_base_reg_block extends uvm_reg_block;
...
+      if (mapped_addr_ranges[0].start_addr - default_map.get_base_addr() > 0) begin
+        range.start_addr = 0;
+        range.end_addr   = mapped_addr_ranges[0].start_addr - 1;
+        unmapped_addr_ranges.push_back(range);
+      end
+
+      // all address ranges in the "middle" - only applies if
+      // there are more than 1 mapped address ranges
+      if (mapped_addr_ranges.size() > 1) begin
+        for (int i = 0; i < mapped_addr_ranges.size() - 1; i++) begin
```
Comment (weicaiyang) [inline hw/dv/sv/dv_base_reg/dv_base_reg_block.sv:176]: this assume mapped_addr_ranges is in order. and we assume mem is located after CSR. if the IP contains 2 reg block, each block has both csr and mem, then the order should be block_1_csr, block_1_mem, block_2_csr, block_2_mem.

> mapped_addr_ranges = {mapped_addr_ranges, mem_ranges};

File: hw/dv/sv/dv_base_reg/dv_base_reg_block.sv [hunk 12]
```diff
@@ -94,6 +103,95 @@ class dv_base_reg_block extends uvm_reg_block;
...
+    // - space between 0 and the first mapped address range (if exists)
+    // - space between the last mapped address and the highest address mapped by the address mask
+    if (mapped_addr_ranges.size() == 0) begin
+      range.start_addr = 0;
+      range.end_addr = get_max_offset();
+      unmapped_addr_ranges.push_back(range);
+    end else begin
+      // 0 -> start_addr-1
+      if (mapped_addr_ranges[0].start_addr - default_map.get_base_addr() > 0) begin
+        range.start_addr = 0;
```
[the remark this answers is not in the document]
Comment (udinator) [reply hw/dv/sv/dv_base_reg/dv_base_reg_block.sv:168]: ohh yes you're right, thanks for catching that one.

File: hw/dv/sv/dv_base_reg/dv_base_reg_block.sv (hunk 10 above)
Comment (matutem) [reply hw/dv/sv/dv_base_reg/dv_base_reg_block.sv:193]: Let profiling determine if we need to optimize this :)

**Review by matutem — APPROVED** [review]
LGTM

File: hw/dv/sv/dv_base_reg/dv_base_reg_block.sv (hunk 10 above)
Comment (udinator) [reply hw/dv/sv/dv_base_reg/dv_base_reg_block.sv:193]: @weicaiyang I believe I implemented it this way just so it would be a little more immediately obvious what the unmapped calculations are trying to do - I think your suggestions make sense also, but like Matute said I feel there's no immediate need to optimize this further unless there's an obvious performance impact, wdyt?

File: hw/dv/sv/dv_base_reg/dv_base_reg_block.sv (hunk 11 above)
Comment (udinator) [reply hw/dv/sv/dv_base_reg/dv_base_reg_block.sv:176]: yes, I agree that this assumption needs to hold in order for this code to work - I believe this is the standard way of mapping out the regmodel for all the IPs though (CSRs first, followed by any memories), so not sure if there are some scenarios where this assumption will not hold true

File: hw/dv/sv/dv_base_reg/dv_base_reg_block.sv (hunk 10 above)
Comment (weicaiyang) [reply hw/dv/sv/dv_base_reg/dv_base_reg_block.sv:193]: Agree. we can optimize it later.

File: hw/dv/sv/dv_base_reg/dv_base_reg_block.sv (hunk 11 above)
Comment (weicaiyang) [reply hw/dv/sv/dv_base_reg/dv_base_reg_block.sv:176]: If it's block level, I feel it's probably yes. but for chip-level, it probably not, as memories are located after all the CSRs in all IPs.

**Review by cindychip — APPROVED** [review]

File: hw/dv/sv/dv_base_reg/dv_base_reg_block.sv (hunk 11 above)
Comment (udinator) [reply hw/dv/sv/dv_base_reg/dv_base_reg_block.sv:176]: good point. maybe I need to change this logic to do an explicit sorting of the address ranges based on the start address then.

Comment (udinator) [reply hw/dv/sv/dv_base_reg/dv_base_reg_block.sv:179]: updated - we now sort the `mapped_addr_ranges` queue in ascending order on the start address of each range.

**Review by weicaiyang — APPROVED** [review]
LGTM, thanks for addressing this.
